tests/resolver/test_lib.py

```diff
@@ -0,0 +1,389 @@
+):
+    # vprops = [
+    #     VariantProperty(namespace="OmniCorp", feature="access_key", value="value1"),
+    #     VariantProperty(
+    #         namespace="TyrellCorporation", feature="client_id", value="value2"
+    #     ),
+    #     VariantProperty(
+    #         namespace="TyrellCorporation", feature="client_pass", value="abcde"
+    #     ),
+    #     VariantProperty(
+    #         namespace="TyrellCorporation", feature="client_pass", value="efghij"
+    #     ),
+    # ]
+
+    # vdescs = [
+    #     # prop3 and prop4 are mutually exclusive
+    #     [00] VariantDescription([prop1, prop2, prop3]),  # Rank 0
+    #     [01] VariantDescription([prop1, prop3, prop2]),  # duplicate
+    #     [02] VariantDescription([prop2, prop3, prop1]),  # duplicate
+    #     [03] VariantDescription([prop2, prop1, prop3]),  # duplicate
+    #     [04] VariantDescription([prop3, prop2, prop1]),  # duplicate
+    #     [05] VariantDescription([prop3, prop1, prop2]),  # duplicate
+    #     [06] VariantDescription([prop1, prop2, prop4]),  # rank 2
+    #     [07] VariantDescription([prop1, prop4, prop2]),  # duplicate
+    #     [08] VariantDescription([prop2, prop4, prop1]),  # duplicate
+    #     [09] VariantDescription([prop2, prop1, prop4]),  # duplicate
+    #     [10] VariantDescription([prop4, prop2, prop1]),  # duplicate
+    #     [11] VariantDescription([prop4, prop1, prop2]),  # duplicate
+    #     [12] VariantDescription([prop1, prop2]),         # Rank 4
+    #     [13] VariantDescription([prop2, prop1]),         # duplicate
+    #     [14] VariantDescription([prop1, prop3]),         # Rank 6
+    #     [15] VariantDescription([prop3, prop1]),         # duplicate
+    #     [16] VariantDescription([prop2, prop3]),         # Rank 1
+    #     [17] VariantDescription([prop3, prop2]),         # duplicate
+    #     [18] VariantDescription([prop1, prop4]),         # Rank 8
+    #     [19] VariantDescription([prop4, prop1]),         # duplicate
+    #     [20] VariantDescription([prop2, prop4]),         # Rank 3
+    #     [21] VariantDescription([prop4, prop2]),         # duplicate
+    #     [22] VariantDescription([prop1]),                # Rank 10
+    #     [23] VariantDescription([prop2]),                # Rank 5
+    #     [24] VariantDescription([prop3]),                # Rank 7
+    #     [25] VariantDescription([prop4]),                # Rank 9
+    # ]
+    vprop1 = VariantProperty(
+        namespace="TyrellCorporation", feature="client_id", value="value2"
+    )
+    vfeat1 = VariantFeature(namespace="TyrellCorporation", feature="client_pass")
+
+    assert sort_and_filter_supported_variants(
+        vdescs=vdescs,
+        supported_vprops=vprops,
+        property_priorities=[vprop1],
+        feature_priorities=[vfeat1],
+        namespace_priorities=["OmniCorp", "TyrellCorporation"],
+    ) == [
+        # everything that contains `vprop2` with vfeat1 important and OmniCorp first
+        vdescs[0],  # 0
```

Comment:
To be honest, this is quite unreadable. Perhaps inlining the `VariantDescription`s here would be a better idea?

Reply:
I can do that :)